updater.go

```diff
@@ -126,23 +130,31 @@ func (u *Updater) update(ctx Context, options UpdateOptions) (*Update, *Error) {
 		return update, verifyErrPtr(err)
 	}
 
-	if err := ctx.BeforeApply(*update); err != nil {
-		return update, applyErrPtr(err)
+	if err := u.apply(ctx, *update, options); err != nil {
+		return update, err
 	}
 
-	if err := u.platformApplyUpdate(*update, options); err != nil {
-		return update, applyErrPtr(err)
+	if err := ctx.Restart(); err != nil {
+		return update, restartErrPtr(err)
```

Comment:
Yes, I think you're exactly right.  I too resisted type-casts, since in C++, it's a sign of weakness. But I think Go wants to you to do it this way. @jzila do you concur?